lib/bolt/fiber_executor.rb

```diff
@@ -0,0 +1,168 @@
+      @id += 1
+      Bolt::PlanFuture.new(future, @id, name)
+    end
+
+    # Register a PlanFuture with the FiberExecutor to be executed while round_robining.
+    def add_future(future)
+      # This is a little incongruous since we created the future in
+      # `create_future`, but a future can't do anything until it's added to
+      # plan_futures, and all futures (main plan and fiber) flow through the
+      # method.
+      msg = "Created future #{future.name}"
+      @logger.trace(msg)
+
+      plan_futures << future
+    end
+
+    # Visit each PlanFuture registered with the FiberExecutor and resume it.
+    # Fibers will yield themselves back, either if they kicked off a
+    # long-running process or if the current long-running process hasn't
+    # completed. If the Fiber finishes after being resumed, store the result in
+    # the PlanFuture and remove the PlanFuture from the FiberExecutor.
+    def round_robin
+      plan_futures.each do |future|
+        # If the Fiber is still running and can be resumed, then resume it
+        @logger.trace("Checking future '#{future.name}'")
+        if future.alive?
+          @logger.trace("Resuming future '#{future.name}'")
+          future.resume
+        end
+
+        # Once we've restarted the Fiber, check to see if it's finished again
+        # and cleanup if it has.
+        next if future.alive?
+        @logger.trace("Cleaning up future '#{future.name}'")
+
+        # If the future errored and the main plan has already exited, log the
+        # error at warn level.
+        unless plan_futures.map(&:id).include?(0) || future.state == "done"
+          Bolt::Logger.warn('errored-futures', "Error in future '#{future.name}': #{future.value}")
+        end
+
+        # Remove the PlanFuture from the FiberExecutor.
+        plan_futures.delete(future)
+      end
+
+      # If the Fiber immediately returned wait for long enough that something
+      # can execute before checking again. This mitigates CPU thrashing.
+      return unless plan_futures.all? { |f| f.value.is_a?(Symbol) && f.value == :returned_immediately }
+      @logger.trace("Nothing can be resumed. Rechecking in 0.5 seconds.")
+
+      sleep(0.5)
+    end
+
+    # Whether all PlanFutures have finished executing, indicating that the
+    # entire plan (main plan and any PlanFutures it spawned) has finished and
+    # Bolt can exit.
+    def plan_complete?
+      plan_futures.empty?
+    end
+
+    # Block until the provided PlanFuture objects have finished, or the timeout is reached.
+    #
+    def wait(futures, timeout, catch_errors: false, **_kwargs)
+      if timeout
+        start = Time.now
+        Fiber.yield(:unfinished) until (Time.now - start > timeout) || futures.map(&:alive?).none?
+        # Raise an error for any futures that are still alive
+        futures.each do |f|
+          if f.alive?
+            f.raise(Bolt::FutureTimeoutError.new(f.name, timeout))
```

Comment:
Fibers aren't returning this exception as their values and are instead returning symbols (specifically `:returned_immediately`). This causes the `wait` function to error, as the return type is not a valid `PlanResult`, and is instead a `RunTime` object.

Example plan:

```puppet
plan bolt(
  TargetSpec $targets = 'localhost'
) {
  $futures = background() || {
    run_command('sleep 5', $targets)
  }

  return wait($futures, 2)
}
```

And the error:

```json
{
  "kind": "bolt/pal-error",
  "msg": "value returned from function 'wait' has wrong type, index 0 expects a Boltlib::PlanResult = Optional[Variant[Boolean, Numeric, String, Object[{name => 'Error', type_parameters => {'kind' => Variant[String, Regexp, Type[Enum], Type[Pattern], Type[NotUndef], Type[Undef]], 'issue_code' => Variant[String, Regexp, Type[Enum], Type[Pattern], Type[NotUndef], Type[Undef]]}, attributes => {'msg' => String[1], 'kind' => {type => Optional[String[1]], value => undef}, 'details' => {type => Optional[Hash[String[1], Data]], value => undef}, 'issue_code' => {type => Optional[String[1]], value => undef}}, functions => {'message' => Callable[[0, 0], String[1]]}}], Object[{name => 'Result', attributes => {'value' => Hash[String[1], Data], 'target' => Target}, functions => {'error' => Callable[[0, 0], Optional[Error]], 'message' => Callable[[0, 0], Optional[String]], 'sensitive' => Callable[[0, 0], Optional[Sensitive[Data]]], 'action' => Callable[[0, 0], String], 'status' => Callable[[0, 0], String], 'to_data' => Callable[[0, 0], Hash], 'ok' => Callable[[0, 0], Boolean], '[]' => Callable[[String[1]], Variant[Data, Sensitive[Data]]]}}], Object[{name => 'ApplyResult', attributes => {'report' => Hash[String[1], Data], 'target' => Target}, functions => {'error' => Callable[[0, 0], Optional[Error]], 'ok' => Callable[[0, 0], Boolean], 'message' => Callable[[0, 0], Optional[String]], 'action' => Callable[[0, 0], String], 'to_data' => Callable[[0, 0], Hash], 'value' => Callable[[0, 0], Hash]}}], Object[{name => 'ResultSet', attributes => {'results' => Array[Variant[Result, ApplyResult]]}, functions => {'count' => Callable[[0, 0], Integer], 'empty' => Callable[[0, 0], Boolean], 'error_set' => Callable[[0, 0], ResultSet], 'filter_set' => Callable[[0, 0, Callable], ResultSet], 'find' => Callable[[String[1]], Optional[Variant[Result, ApplyResult]]], 'first' => Callable[[0, 0], Optional[Variant[Result, ApplyResult]]], 'names' => Callable[[0, 0], Array[String[1]]], 'ok' => Callable[[0, 0], Boolean], 'ok_set' => Callable[[0, 0], ResultSet], 'targets' => Callable[[0, 0], Array[Target]], 'to_data' => Callable[[0, 0], Array[Hash]], '[]' => Variant[Callable[[Integer], Optional[Variant[Result, ApplyResult, Array[Variant[Result, ApplyResult]]]]], Callable[[Integer, Integer], Optional[Variant[Result, ApplyResult, Array[Variant[Result, ApplyResult]]]]]]}}], Object[{name => 'Target', attributes => {'uri' => {type => Optional[String[1]], kind => given_or_derived}, 'name' => {type => Optional[String[1]], kind => given_or_derived}, 'safe_name' => {type => Optional[String[1]], kind => given_or_derived}, 'target_alias' => {type => Optional[Variant[String[1], Array[String[1]]]], kind => given_or_derived}, 'config' => {type => Optional[Hash[String[1], Data]], kind => given_or_derived}, 'vars' => {type => Optional[Hash[String[1], Data]], kind => given_or_derived}, 'facts' => {type => Optional[Hash[String[1], Data]], kind => given_or_derived}, 'features' => {type => Optional[Array[String[1]]], kind => given_or_derived}, 'plugin_hooks' => {type => Optional[Hash[String[1], Data]], kind => given_or_derived}, 'resources' => {type => Optional[Hash[String[1], ResourceInstance]], kind => given_or_derived}}, functions => {'host' => Callable[[0, 0], Optional[String]], 'password' => Callable[[0, 0], Optional[String[1]]], 'port' => Callable[[0, 0], Optional[Integer]], 'protocol' => Callable[[0, 0], Optional[String[1]]], 'transport' => Callable[[0, 0], String[1]], 'transport_config' => Callable[[0, 0], Hash[String[1], Data]], 'user' => Callable[[0, 0], Optional[String[1]]]}}], Object[{name => 'ResourceInstance', attributes => {'target' => Target, 'type' => Variant[String[1], Type[Resource]], 'title' => String[1], 'state' => {type => Optional[Hash[String[1], Data]], value => undef}, 'desired_state' => {type => Optional[Hash[String[1], Data]], value => undef}, 'events' => {type => Optional[Array[Hash[String[1], Data]]], value => undef}}, functions => {'add_event' => Callable[[Hash[String[1], Data]], Array[Hash[String[1], Data]]], 'set_state' => Callable[[Hash[String[1], Data]], Hash[String[1], Data]], 'overwrite_state' => Callable[[Hash[String[1], Data]], Hash[String[1], Data]], 'set_desired_state' => Callable[[Hash[String[1], Data]], Hash[String[1], Data]], 'overwrite_desired_state' => Callable[[Hash[String[1], Data]], Hash[String[1], Data]], 'reference' => Callable[[0, 0], String], '[]' => Callable[[String[1]], Data]}}], Object[{name => 'ContainerResult', attributes => {'value' => Hash[String[1], Data]}, functions => {'[]' => Callable[[String[1]], Data], 'error' => Callable[[0, 0], Optional[Error]], 'ok' => Callable[[0, 0], Boolean], 'status' => Callable[[0, 0], String], 'stdout' => Callable[[0, 0], String], 'stderr' => Callable[[0, 0], String], 'to_data' => Callable[[0, 0], Hash]}}], Object[{name => 'Future', functions => {'state' => Callable[[0, 0], Enum['done', 'error', 'running']]}}], Array[Boltlib::PlanResult], Hash[String, Boltlib::PlanResult]]] value, got Runtime",
  "details": {
    "file": "/Users/tom/.puppetlabs/bolt/plans/init.pp",
    "line": 8,
    "column": 10
  }
}
```

Reply:
Oh my gosh, `Fiber.raise` wasn't added until Ruby 2.7 https://rubyreferences.github.io/rubychanges/2.7.html#fiberraise ughghghghghhg

Reply:
Ah, I missed setting the `value` of the PlanFuture when raising errors. Nice!